api/v1alpha1/istiorevisiontags_types.go

```diff
@@ -0,0 +1,199 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	IstioRevisionTagKind = "IstioRevisionTag"
+	DefaultRevisionTag   = "default"
+)
+
+// IstioRevisionTagSpec defines the desired state of IstioRevisionTag
+type IstioRevisionTagSpec struct {
+	// +kubebuilder:validation:Required
+	TargetRef IstioRevisionTagTargetReference `json:"targetRef"`
+}
+
+// IstioRevisionTagTargetReference can reference either Istio or IstioRevision objects in the cluster.
+type IstioRevisionTagTargetReference struct {
+	// Kind is the kind of the target resource.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Kind string `json:"kind"`
+
+	// Name is the name of the target resource.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
+}
+
+// IstioRevisionStatus defines the observed state of IstioRevision
+type IstioRevisionTagStatus struct {
+	// ObservedGeneration is the most recent generation observed for this
+	// IstioRevisionTag object. It corresponds to the object's generation, which is
+	// updated on mutation by the API Server. The information in the status
+	// pertains to this particular generation of the object.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Represents the latest available observations of the object's current state.
+	Conditions []IstioRevisionTagCondition `json:"conditions,omitempty"`
+
+	// Reports the current state of the object.
+	State IstioRevisionTagConditionReason `json:"state,omitempty"`
+
+	// IstiodNamespace stores the namespace of the corresponding Istiod instance
+	IstiodNamespace string `json:"istiodNamespace"`
+
+	// IstioRevision stores the name of the referenced IstioRevision
+	IstioRevision string `json:"istioRevision"`
```

Comment:
no, actually the reasoning behind this field is exactly the fact that we could be referencing an Istio resource. The IstioRevision controller needs some easy way to know which IstioRevision is being referenced (even if technically it is referencing an Istio, it will always resolve to an IstioRevision), so instead of duplicating the resolution logic behind that, I'm just doing it at reconcile time in the IstioRevisionTag controller and write the result into this field